Display plugin inspections in dependencies output

dgmstuart opened the pull request:
_(replaces https://github.com/dxw/whippet/pull/92 - identical except that it uses mockery instead of phpunit 5.4)_

When running `whippet deps install` or `whippet deps update` (which
does the install step as well as updating), output a link to any inspections
which dxw have done for that plugin and published on https://security.dxw.com/

Example output:

```
[Checking plugins/akismet]
HEAD is now at 9ca38d7... v3.2
Inspections for this plugin:
* 13/07/2016 - 1.2.3 - No issues found - http://security.dxw.com/plugins/akismet/

[Checking plugins/advanced-custom-fields-pro]
HEAD is now at 6fa6118... advanced-custom-fields-pro-5.4.8.zip
#############################################
#                                           #
#  WARNING: No inspections for this plugin  #
#                                           #
#############################################
```

Merge Request for the API: https://git.govpress.com/dxw-advisories/app/merge_requests/15

## How to test

1. pull down the dxw security project on the MR branch (https://git.govpress.com/dxw-advisories/app/merge_requests/15)
2. get a database dump from production from ops, and install it (so that you have some inspections)
3. run the dxw security site locally
4. in your local whippet find`src/Modules/Dependencies.php` and change this line to point at the host that the dxw-advisories site is running on, (e.g. 'http://localhost:8000') 

    ```php
    $inspections_api_host = 'https://security.dxw.com';
    ```
5. run `whippet deps install` on a project with a few plugins

## TODO

Three tasks remain - I think these can be deferred to further PRs:

- DONE ~~Including the inspection result in the output. This currently
  doesn't come back from the API, but it should be fairly easy to add it.~~
- DONE ~~Including error messages in output. Currently if an error arises while
  trying to retrieve and display an inspection, nothing is output.
  I considered stopping the installation running, but I don't think
  we want to stop whippet running e.g. if the API is down!~~
- DONE - search is now done via a new 'slug' custom field ~~Displaying multiple inspections. Currently if multiple inspections have
  been carried out over time, these have urls with incrementing numbers
  on the end, e.g. https://security.dxw.com/plugins/the-events-calendar-pro-2/
  One way of doing this would be to make multiple requests to the API: if you
  find one, keep incrementing the number until you can't find anymore.~~

## Note
This was test-driven in the 'mockist' style of TDD:

1. start from somewhere near the top and assume the existence of collaborating
  objects (in this case all dependency-injected).
2. Mock out those collaborators in order to get the tests to your object to pass
3. Repeat all the way down until you reach the boundary of the system.

dgmstuart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add a feature to whippet to query dxwsec for inspection results for installed plugins, and prompt when one needs a new inspection. Possibly use wp-json for the API?](https://trello.com/c/L2Wi4Qh8/10-add-a-feature-to-whippet-to-query-dxwsec-for-inspection-results-for-installed-plugins-and-prompt-when-one-needs-a-new-inspection)

dgmstuart commented:
@tomdxw This is ready for a look I think

RobjS commented on src/Services/JsonApi.php:
```diff
@@ -0,0 +1,31 @@
+
+        $response_as_array = json_decode($response->unwrap(), true);
+        if (is_null($response_as_array)) {
+            return \Result\Result::err('Received invalid JSON when requesting '.$url);
```
Would it be useful to know the response code when you get an error? Then if/when things go wrong, it'd be easier to work out what was happening.

dgmstuart replied:
OK, I've switched to Guzzle and differentiated connection errors from application errors. Hard to guess what kind of errors we might get: I suggest we wait for real errors to happen and add cases handle them as we go - should be pretty easy in the new setup.

dgmstuart replied:
I've clarified this error message to say that we couldn't parse the JSON, and also added some code to check the JSON response, so we'll get errors if the JSON is not as expected.

RobjS commented on src/Services/InspectionsApi.php:
```diff
@@ -0,0 +1,59 @@
+
+    private function url($plugin_slug)
+    {
+        return $this->host.'/wp-json/v1/inspections/'.$plugin_slug;
```
I'd be tempted to remove all the url-related variables (both here and the `$inspections_api_host` defined in Dependencies.php), and define them all in one place - preferably a config file of some sort (maybe whippet.json?). That'd make them easier to manage, and avoid the risk of tests/functionality being broken by a url change.

dgmstuart replied:
I went a fair way down this path, but it seems like whippet.json is strongly associated with dependencies: the hash of that file is used to determine whether the lockfile needs to be updated, so changing the API host or Path results in the whole lock file needing to be updated, which feels wrong to me.

What I _have_ done is move the _path_ definition to the same place as the host. 

Although it would be best to manage these URLs as config, I feel like they are going to be fairly stable, so could we consider shipping this as-is?

RobjS commented on src/Services/InspectionsApi.php:
```diff
@@ -0,0 +1,59 @@
+    }
+}
+
+class Inspection
```
Hmm, I can see the logic behind this class being in the InspectionsApi.php file, seeing as `InspectionsApi` is the only place where it's used, but multiple classes in a single file makes me uneasy. If I had a reason to be specifically looking for this class, I wouldn't really know where to start.

dgmstuart replied:
Yup - agreed. Would have been nice to use some sort of Struct object inline, but that doesn't seem to exist in PHP. 

There's nowhere in the current directory structure for this to exist (it isn't in itself a Service object), so I've created a new 'Models' directory.

dgmstuart commented:
@RobjS Ready for another look I think. Unfortunately probably shouldn't be merged until https://git.govpress.com/dxw-advisories/app/merge_requests/17 is merged? Though actually it shouldn't matter too much: people will just get "couldn't connect to api" errors when updating deps, which might encourage us to get the dxw-advisories MR merged 😜

dgmstuart commented:
Fixed up: replaced "PSR-4" with "PSR-2" in commit messages

RobjS commented on src/Dependencies/Installer.php:
```diff
@@ -27,6 +29,17 @@ public function install()
+                    $error = $result->getErr();
+                    echo sprintf("[ERROR] %s\n", $error);
+                }
+                echo "\n";
```
I wonder if it'd be worth extracting the `inspectionChecker` stuff (lines 32-42) to a separate method within `Installer`? Then you could just call that from within `install()`, by doing something like `$this->displayInspections($type, dep);`. It'd make it easier to see at a glance what the `install()` method does.

dgmstuart replied:
Good shout - I've extracted a method and rebased.

dgmstuart commented:
Ready for another look.

RobjS commented on src/Dependencies/Installer.php:
```diff
@@ -96,4 +87,22 @@ private function installDependency($type, $dep)
+        if (!$result->isErr()) {
+            $inspectionDetails = $result->unwrap();
+            if (!empty($inspectionDetails)) {
+                $message = sprintf("%s\n", $inspectionDetails);
```
If you really want to avoid the nested conditionals, you could do something like:

```
$message = null;
if (!$result ->isErr() && !empty($result->unwrap())) {
    $message = sprintf("%s\n", $result->unwrap());
} else {
    $error = $result->getErr();
    $message = sprintf("[ERROR] %s\n", $error);
}
return $message;
```

But I'm not really convinced that's any better, so I'll leave it up to you.

dgmstuart replied:
I don't think that works: if `empty($result->unwrap())` then you'll end up in the `else` branch....

dgmstuart replied:
You'd need that to be `else if($result ->isErr()) {`

RobjS replied:
Oops, yeah. 

You could probably do without the `else if`:

```
$message = null;
if (!$result ->isErr() && !empty($result->unwrap())) {
...
}
if ($result->isErr()) {
...
}
return $message;
```

dgmstuart replied:
Sure. At this point we have a variable being set and overriden, in theory multiple times (until you mentally parse that those two `if`s are mutually exclusive. This sort of situation leads to code which is hard to debug, since there are several places that a variable might be getting its value from. 

The nested conditional is bad, but at least it's clear that `$message` is getting its value from one of three places.

dgmstuart commented:
Rebased. Had to resolve some conflicts with https://github.com/dxw/whippet/pull/95

dgmstuart commented:
Rebased, and added a couple of features

- disable the inspection checks via a command line parameter
- allow the API host to be chosen via an environment variable

tomdxw commented:
> Rob has approved everything up to the last two commits

And the last two commits look fine.
